docs: Language Constraints and Upgrading (#28856)
renovatebot/renovate: docs/usage/language-constraints-upgrading.md

For certain ecosystems, changes to language constraints are usually done in major releases and documented in the release notes, so this behavior may not be problematic.
For other ecosystems this may seem simply _wrong_.

Renovate users can opt into strict compatibility filtering using `constraintsFiltering=strict`.
You should understand the limitations of this first, also as an understanding of why it's not the default behavior.

## Language constraint updating

The first challenge is that Renovate may not yet support the ability to update your language constraints in an automated manner, and even when it does, users may not understand how many updates are depending on it.

For example, consider a Node.js project which has its `engines` set to `"node": "^18.0.0 || ^20.0.0"`.

Should Renovate skip Node.js v21 because it's non-LTS?
When Node.js v22 comes out, should Renovate add it to your engines, or wait until it starts LTS?
When Node.js v18 is EOL, should Renovate drop it from the engines list?

It's hard for Renovate to know or guess what users want, because users have strong but also different opinions on the answer to the above questions.

> Additionally, even if Renovate guesses right or adds advanced capabilities to allow this to be configurable, then users might sit any of these "major" upgrades for months.

Reviewer: ```suggestion
Also, even _if_ Renovate guesses right or adds advanced capabilities to allow this to be configurable: users might still wait on any of these "major" upgrades for months.
```

> If a project delays creating or merging the update to drop Node.js v18 from engines then it won't be able to upgrade to any new versions of library dependencies which themselves already dropped support.

Reviewer: ```suggestion
If a project waits to create or merge the update to drop Node.js `v18` from `engines`, then they can _not_ upgrade to any new versions of library dependencies.
Those library dependencies may have dropped support for Node.js `v18` already.
```

## Strict filtering limitations

> Consider again the Node.js project which has its `engines` set to `"node": "^18.0.0 || ^20.0.0"`.

Reviewer: ```suggestion
Let's go back to the Node.js project which has its `engines` field set to `"node": "^18.0.0 || ^20.0.0"`.
```

> Now also consider a library which sets its `engines` to `"node": "^18.12.0 || ^20.9.0"` because it wants to only support "LTS releases" of Node.js.

Reviewer: ```suggestion
Now also consider a library which sets its `engines` field to `"node": "^18.12.0 || ^20.9.0"` because the library only supports "LTS releases" of Node.js.
```

> Strictly speaking, this dependency is not compatible with the project above which has wider requirements for Node versions, and Renovate would hold back any upgrades for it.

Reviewer: ```suggestion
Strictly speaking, this library is _not_ compatible with the project above, because the project has _wider requirements_ for their Node versions.
This means Renovate holds back any upgrades for it.
```

> Should Renovate somehow reason and _assume_ that this narrower engines support is actually OK?

Reviewer: ```suggestion
Should Renovate somehow "think" and _assume_ that this narrower `engines` support is actually OK?
```

> What if the project _already_ used an existing version of this library "incorrectly"?

Reviewer: ```suggestion
What if the project _already_ used a current version of this library "in a way that's not officially supported"?
```

> A second problem is that if Renovate is unable to update the language constraints, or the user ignores or does not notice the language upgrade, then they also may not realize that many of their dependencies are out of date because Renovate is not proposing PRs.

Reviewer: ```suggestion
A second problem is that if:

- Renovate can _not_ update the language constraints, or
- a user _ignores_ or does not see the language upgrade

Then the user may not know that many dependencies are out of date, because Renovate is not creating PRs.
```

> For example, a project may have 10 dependencies, and 8 of those have updates but all 8 require the project to update its language constraints first.

Reviewer: ```suggestion
For example: a project may have 10 dependencies, and 8 of those have updates.
But all 8 dependencies need the project to update its language constraints _first_.
```

> The project administrator may think they are mostly up to date, because Renovate is not proposing PRs, but in reality 80% of their dependencies are outdated.

Reviewer: ```suggestion
The project administrator thinks they are up to date, because Renovate is not creating PRs, but 80% of their dependencies are outdated.
```

> In short, users who apply `constraintsFiltering=strict` usually understimate just how strict that will be and how many releases will be filtered out.

Reviewer: ```suggestion
In short, users who set `constraintsFiltering=strict` often do not understand how _strict_ that setting is and how many releases it will _filter out_.
```

## Transitive constraint limitations

It is a common problem that a library may declare language constraints (such as the `engines` from the above examples) but then itself depend on one or more libraries which have narrower contraints (such as `"node": "^20.0.0"`).

> In such scenarios Renovate "trusts" the declaration of the library and may propose it as an update even with strict constraints filtering.

Reviewer: ```suggestion
In cases like these, Renovate "trusts" the declaration of the library and may create a update, even _with_ strict constraints filtering.
```

> For some package managers, such as npm, this incompatibility won't be detected or warned (even during lock file generation), although it may also not be a problem in your application either.

Reviewer: ```suggestion
For some package managers, like `npm`, this incompatibility will _not_ be detected or warned about (even during lock file generation), but this may not be a problem for your application.
```

> Other package managers, such as Poetry, may detect incompatible language constraints during lock file generation and Renovate will report that as an "Artifacts update error".

Reviewer: ```suggestion
Other package managers, like Poetry, may detect and warn about incompatible language constraints during lock file generation, which Renovate reports as an "Artifacts update error".
```

## Applying constraints through config

> Renovate supports the manual configuration of language constraints in its config. For example:

Reviewer: ```suggestion
You can set language constraints in the Renovate config.
For example:
```

```json title="Renovate config with Node.js constraints"
{
  "constraints": {
    "node": "^18.0.0 || >=20.0.0"
  }
}
```

> Defining constraints manually using Renovate config can be necessary for the following use cases:

Reviewer: ```suggestion
You may need to set constraints in the Renovate config when:
```

> - The package manager of the project doesn't support constraints declarations, or

Reviewer: ```suggestion
- The package manager of the project does not support constraints declarations, or
```

> - The project hassn't declared any constraints, or

Reviewer: ```suggestion
- The project has not declared any constraints, or
```

> - The user wants Renovate to use different constraints to what's declared in the project.

Reviewer: ```suggestion
- You want Renovate to use _different_ constraints to what's declared in the _project_
```

> Renovate will not propose "update" PRs proposing to update any of these versions once they become outdated, so the user is responsible for keeping them updated manually.

Reviewer: ```suggestion
Renovate will _not_ create "update" PRs to update any of these versions once they become outdated, so you must update those by hand.
```

> For this reason, configuring constraints manually in config is considered _undesirable_ and we'd prefer to fix problems in Renovate instead if that's the reason for needing it.

Reviewer: ```suggestion
For this reason, setting constraints manually in the Renovate config is _undesirable_.
We prefer to fix problems in Renovate itself, instead of you setting constraints.
```

## Future Work

> Please start or join a GitHub Discussion if you have an interest in this topic.

Reviewer: ```suggestion
Please start, or join, a GitHub Discussion if you are interested in this topic.
```
